### DISCUSSION TITLE
Type Conversion

### DISCUSSION THREAD

**Discussion by panarch** [gluesql/gluesql#147] [Ideas]
Let's talk about type conversion in here.

started in https://github.com/gluesql/gluesql/pull/143 thread.
![image](https://user-images.githubusercontent.com/2025065/109383304-1782cc00-7929-11eb-8a76-8ff26439666f.png)

### I think implicit conversion between number types are good, but not sure about between string and number types.

In the case of number types, most of cases we can safely convert types without panic.
(usually we may use conversion to float types like `f64`)

However, string types are different.
Implicit conversion between string and number can cause some problems when string values cannot be parsed so we need to use something like `NaN`.
Also in some cases, it would be a bit hard to decide convert whether from string to number or from number to string.

I'm not sure that specifying a straightforward policy of implicit conversion between number and string is really possible.
We all know an example of... javascript, about why we don't use `==` and `!=` anymore.

**Comment by KyGost** [comment]
What if we throw an error instead of `NaN`?

Also conversion would be for conforming to table types so wouldn't need to decide which way to convert.

I work with MSSQLServer a bunch for work and there it tries to convert but if an given query it runs into a single case where it can't comfortably convert it just throws like `Could not convert [value] to type [type]` or something like that.

**Reply by panarch** [reply]
Good point, then `NaN` may not be an issue anymore.

However, still the ambiguity issue remains.
`1 + 1 => 2` and `"1" + "1" => "11"` are straightforward, but... how about `"1" + 1` or `1 + "1"`? There can be two answers, `2` or `"11"`.

**Reply by KyGost** [reply]
Hmmm I see your point.
MSSQL seems to have an order of preference for conversion or something; `'1' + 1 = 1 + '1' = 2` (they all evaluate to ints).

**Reply by KyGost** [reply]
I think order of preference would work pretty well.
I don't think it's unreasonable to assume that any time a string and an int are encountered one assumes int is desired.
nb: MSSQL throws an error for `'A' + 1`

**Reply by panarch** [reply]
haha, good to see that MSSQL evaluate `'1' + 1 = 2`
Because..... `'1' + 1 = 1 + '1' = '11'` in javascript.

MSSQL seems like having consistent behavior, only trying to convert into number types.

I just tested using Sqlite3,
* `SELECT 1 + "1";` => 2
* `1.0 = "1.0"` => false

**Reply by KyGost** [reply]
Hahaha JavaScript can be a mess.

Maybe there's general SQL documenation out there somewhere that specifies?

Strange.... In MSSQL:
`SELECT 1 WHERE 1.0 = '1.0'`
->`1`

**Reply by panarch** [reply]
Oow... may be ANSI SQL 2016 has something..?
hmm.. there looks quite a several different implementations.

**Comment by panarch** [comment]
Ok, let's divide the type conversion issue.
* implicit type conversion between number types (for now, float & integer)
* implicit type conversion between string & number

Because... currently we even don't support conversion between number types. (only conversion between literal (`AstValue`) is implemented)

There's no doubt that we need to properly support number type conversion.
How about starting from it?

**Reply by KyGost** [reply]
That sounds like a good starting point.
Will work on after #144 is merged.
